ros2cli/test/test_direct.py

```diff
@@ -0,0 +1,63 @@
+# Copyright 2020 Sony Corporation.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import argparse
+
+import pytest
+
+from ros2cli.node.direct import add_arguments as add_direct_node_arguments
+from ros2cli.node.direct import DEFAULT_TIMEOUT
+from ros2cli.node.direct import DirectNode
+
+TEST_NODE_NAME = 'test_node'
+
+
+@pytest.fixture(scope='function')
+def test_arguments_parser():
+    parser = argparse.ArgumentParser()
+    add_direct_node_arguments(parser)
+    return parser
+
+
+def test_default_arguments(test_arguments_parser):
+    args = test_arguments_parser.parse_args()
+    assert not args.use_sim_time
+    assert DEFAULT_TIMEOUT == args.spin_time
+
+
+def test_use_sim_time_arguments(test_arguments_parser):
+    args = test_arguments_parser.parse_args(['--use-sim-time'])
+    assert args.use_sim_time
+    assert DEFAULT_TIMEOUT == args.spin_time
+
+
+def test_spin_time_arguments(test_arguments_parser):
+    args = test_arguments_parser.parse_args(['--spin-time', '10.0'])
+    assert not args.use_sim_time
+    assert 10.0 == args.spin_time
+
+
+def test_use_sim_time_parameter():
+    with DirectNode(args=[], node_name=TEST_NODE_NAME) as direct_node:
+        assert not direct_node.node.get_parameter('use_sim_time').value
+
+    args = argparse.Namespace(use_sim_time=False)
+    with DirectNode(args, node_name=TEST_NODE_NAME) as direct_node:
+        assert not direct_node.node.get_parameter('use_sim_time').value
+
+    # TODO(fujitatomoya): enable this test once /clock callback thread is insulated.
```

Comment:
i did try using threading to run DirectNode in another thread, but the thread cannot be finished anyway. instead of fighting here, it would be nice to address https://github.com/ros2/rclcpp/issues/1542 for rclpy, so that we can have a control for `Node`.

Any idea?

Reply:
Do I understand the problem correctly?: The commented test does not work because the node does not shutdown since it is not getting the simulation clock signal. If the signal was insulated, then the inner logic of the node would allow it to shutdown after the `with` context for the node has been left in the test, which destroys the node.

The suggestion you've linked to might be a good approach. I know that there are active discussions on how to make clocks more consistent in ROS2. My impression is that more discussion needs to occur before decisions are made and implementation can begin. This may mean that the test may be blocked for some time. 

Perhaps we could proceed with another solution that works for now (e.g., kill the node's thread after you've confirmed the value), or leave the TODO to enable the test, as you have done.

Reply:
> Do I understand the problem correctly?

right, node just waits on rmw_wait forever to subscribe /clock topic. test just keeps running and cannot be finished.

> My impression is that more discussion needs to occur before decisions are made and implementation can begin. This may mean that the test may be blocked for some time.

totally agree,

> Perhaps we could proceed with another solution that works for now (e.g., kill the node's thread after you've confirmed the value), or leave the TODO to enable the test, as you have done.

i think we cannot simply kill the threads with python threading, if you know the idea that would be really help. (i really do not want to introduce multi-processing on this test) probably we could send the SIGINT signal to self-pid and join thread?

Reply:
none of them are working to finish the pytest of test_use_sim_time_true_parameter. DirectNode expects timer expire but it just waiting on that event via syscall futex. i'd like to leave this TODO atm.

```
def test_use_sim_time_true_parameter():
    # TODO(fujitatomoya): enable this test once /clock callback thread is insulated.
    #                     see https://github.com/ros2/rclcpp/issues/1542
    direct_node = None

    def spin_direct_node():
        nonlocal direct_node
        args = argparse.Namespace(use_sim_time=True)
        direct_node = DirectNode(args, node_name=TEST_NODE_NAME)

    t = threading.Thread(target=spin_direct_node)
    t.start()
    time.sleep(0.1)
    assert direct_node.node.get_parameter('use_sim_time').value
    #pytest.exit("exit anyway", not direct_node.node.get_parameter('use_sim_time').value)
    #direct_node.node.destroy_timer(direct_node.timer)
    #os.kill(os.getpid(), signal.SIGINT)
    t.join()
```

Reply:
> i'd like to leave this TODO atm.

Sounds good to me.